util/etcd/etcd.go

```diff
@@ -0,0 +1,368 @@
+type Member struct {
+	// ID is the ID of this cluster member
+	ID uint64
+	// Name is the human-readable name of the member. If the member is not started, the name will be an empty string.
+	Name string
+	// PeerURLs is the list of URLs the member exposes to the cluster for communication.
+	PeerURLs []string
+	// ClientURLs is the list of URLs the member exposes to clients for communication. If the member is not started, clientURLs will be empty.
+	ClientURLs []string
+	// IsLearner indicates if the member is raft learner.
+	IsLearner bool
+	// Alarms is the list of alarms for a member
+	Alarms []AlarmType
+}
+
+type Status struct {
+	// Leader is the member ID which the responding member believes is the current leader.
+	Leader uint64
+
+	// Version is the cluster protocol version used by the responding member
+	Version string
+
+	// Errors contains alarm/health information and status.
+	Errors []string
+
+	// IsLearner indicates if the member is raft learner.
+	IsLearner bool
+
+	// RaftIndex is the current raft committed index of the responding member.
+	RaftIndex uint64
+
+	// RaftTerm is the current raft term of the responding member.
+	RaftTerm uint64
+
+	// RaftAppliedIndex is the current raft applied index of the responding member.
+	RaftAppliedIndex uint64
+
+	// DbSizeInUse is the size of the backend database logically in use, in bytes, of the responding member.
+	DbSizeInUse int64
+
+	// DdSize is the size of the backend database physically allocated, in bytes, of the responding member.
+	DbSize int64
+}
+
+// New creates a new etcd client with a custom dialer
+func NewClient(dialer grpcDial, endpoint string, tlsConfig *tls.Config, logger logr.Logger) (*Client, error) {
+	if endpoint == "" {
+		endpoint = "localhost"
+	}
+	etcdClient, err := clientv3.New(clientv3.Config{
+		Endpoints:   []string{endpoint},
+		DialTimeout: etcdTimeout,
+		DialOptions: []grpc.DialOption{
+			grpc.WithBlock(), // block until the underlying connection is up
+			grpc.WithContextDialer(dialer),
+		},
+		TLS: tlsConfig,
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to create etcd client")
+	}
+	c := &Client{
+		timeout:       etcdTimeout,
+		backoffParams: etcdBackoff,
+		endpoint:      endpoint,
+		log:           logger,
+		etcdClient:    etcdClient,
+	}
+	return c, nil
+}
+
+// Members retrieves a list of etcd members
+func (c *Client) Members() (*[]Member, error) {
+	var lastError error
+	var resp *clientv3.MemberListResponse
+	err := wait.ExponentialBackoff(c.backoffParams, func() (bool, error) {
+		ctx, cancel := context.WithTimeout(context.Background(), c.timeout)
+		var err error
+		resp, err = c.etcdClient.MemberList(ctx)
+		cancel()
+		if err == nil {
+			return true, nil
+		}
+		c.log.Error(err, "failed to get list of members for etcd cluster")
+		lastError = err
+		return false, nil
+	})
+	if err != nil {
+		return nil, errors.Wrap(lastError, "failed to get list of members for etcd cluster")
+	}
```

Comment:
This pattern seems to be reused in multiple parts of the code, which suggests we might want to carve it out and make it its own `doRequest` or similar method / function.

Other comments
- Context cancellation should go after the creation in a `defer` statement, see https://golang.org/pkg/context/#WithTimeout
- The Error log inside the exponential backoff might be misleading, we should change it to a higher V info instead and definitely log in the `if err != nil` block after the backoff call
- The variable `lastError` seems unnecessary if we wrap this pattern somewhere else.